### ISSUE TITLE
Generic deriving ToParamSchema fails for a Tuple

### ISSUE THREAD

**Issue by alanz** [GetShopTV/swagger2#61]
```haskell
instance ToParamSchema (Int,Int)
```

fails with

```
        No instance for (Data.Swagger.Internal.ParamSchema.GToParamSchema
                           (M1
                              C
                              GHC.Generics.C1_0(,)
                              (S1 NoSelector (Rec0 Int) :*: S1 NoSelector (Rec0 Int))))
          arising from a use of ‘Data.Swagger.Internal.ParamSchema.$gdmtoParamSchema’
        In the expression:
          Data.Swagger.Internal.ParamSchema.$gdmtoParamSchema
        In an equation for ‘toParamSchema’:
            toParamSchema = Data.Swagger.Internal.ParamSchema.$gdmtoParamSchema
        In the instance declaration for ‘ToParamSchema (Int, Int)’
```
I would expect it to end up something like

```json
  "parameters": [
      {
          "required": true,
          "schema": {
              "items": [
                  {
                      "maximum": 9223372036854775807,
                      "minimum": -9223372036854775808,
                      "type": "integer"
                  },
                  {
                      "maximum": 9223372036854775807,
                      "minimum": -9223372036854775808,
                      "type": "integer"
                  }
              ],
              "type": "array"
          },
          "in": "body",
          "name": "body"
      }
```

**Comment by fizruk** [comment]
@alanz fair enough.

Another solution I think is to provide instances in the library for tuples.
That is, to not change the generic mechanism.
Because if we are changing it then some questions arise:
- should this also work for `data Foo = Bar Int String`?
- should this also work for `data Foo = Bar { x :: Int, y :: String }`?

Personally I am not sure encoding the last one as `[x, y]` would make much sense.

**Comment by alanz** [comment]
I would be happy with an instance for a tuple, as I have not been able to create one myself yet.  And the example I quoted above comes from modifying the example `Todo.hs` in `servant-swagger` examples to have the new route "foo" which takes a tuple param. 

```haskell
type Pos = (Int,Int)


todoAPI :: Proxy TodoAPI
todoAPI = Proxy

-- | The API of a Todo service.
type TodoAPI
    = "todo" :> Get '[JSON] [Todo]
 :<|> "foo"  :> ReqBody '[JSON] Pos :>Get '[JSON] [Todo]
 :<|> "todo" :> ReqBody '[JSON] Todo :> Post '[JSON] TodoId
 :<|> "todo" :> Capture "id" TodoId :> Get '[JSON] Todo
 :<|> "todo" :> Capture "id" TodoId :> ReqBody '[JSON] Todo :> Put '[JSON] TodoId
```

**Comment by alanz** [comment]
@fizruk I am completely stuck here.

Could you give me an example of how to declare

```haskell
instance ToParamSchema (Int,Int)
```
?  The comment for `SwaggerItems` says "SwaggerItemsArray should be used to specify tuple Schemas" but I get a type error for  `SwaggerItems Schema` being provided instead of `SwaggerItems t`.

And an error for overlapping `HasItems` instances.

**Comment by fizruk** [comment]
@alanz hmm, I now see why I didn't have instances for tuples.

Swagger specification does not allow `SwaggerItemsArray` in `ParamSchema`, but only in `Schema`.
So if your `ParamSchema` has type `SwaggerArray` then it has to be homogenous.

In case of a homogenous tuple you can provide a `ParamSchema` like this:

```haskell
{-# LANGUAGE ScopedTypeVariables #-}

instance ToParamSchema a => ToParamSchema (a, a) where
  toParamSchema _ = toParamSchema (Proxy :: Proxy [a])
    & maxItems ?~ 2
    & minItems ?~ 2
```

Since there is no general solution (for heterogenous tuples) I no longer think this should be in the library.

Note that `ParamSchema` should correspond to `ToText` (for `servant < 0.5`) and `ToHttpApiData` (for `servant`'s HEAD). Since there is no default instance for tuples of any of those classes, I suppose you have some custom instance. To build a correct `ParamSchema` it would help if you provide the corresponding `ToText` or `ToHttpApiData` instance.

Another pointer here is that since there is no default instance for tuples of `ToParamSchema`, `ToText`/`FromText` and `ToHttpApiData`/`FromHttpApiData` you probably should not use it for query parameters/captures/headers. But I don't know what logic does `(Int, Int)` represent, so I can't offer you a suitable alternative.

**Comment by alanz** [comment]
Thanks.  FYI (Int,Int) represents a text position line and col, and I think the simplest will be to separate them in hie, which works more cleanly on the API anyway.

But the thing that confuses me is that the `servant-swagger` comes up with a representation which I put in my initial report, and have been using as a goal ever since.

**Comment by fizruk** [comment]
> I think the simplest will be to separate them in hie, which works more cleanly on the API anyway.

That's what I recommend, yes.

> But the thing that confuses me is that the `servant-swagger` comes up with a representation which I put in my initial report, and have been using as a goal ever since.

Can you rephrase/elaborate on this? If something is confusing you, I would love to fix that!

**Comment by alanz** [comment]
What confuses me is that I get

```json
  "parameters": [
      {
          "required": true,
          "schema": {
              "items": [
                  {
                      "maximum": 9223372036854775807,
                      "minimum": -9223372036854775808,
                      "type": "integer"
                  },
                  {
                      "maximum": 9223372036854775807,
                      "minimum": -9223372036854775808,
                      "type": "integer"
                  }
              ],
              "type": "array"
          },
          "in": "body",
          "name": "body"
      }
```

From amending the `servant-swagger` example to have
```haskell
type Pos = (Int,Int)


todoAPI :: Proxy TodoAPI
todoAPI = Proxy

-- | The API of a Todo service.
type TodoAPI
    = "todo" :> Get '[JSON] [Todo]
 :<|> "foo"  :> ReqBody '[JSON] Pos :>Get '[JSON] [Todo]
 :<|> "todo" :> ReqBody '[JSON] Todo :> Post '[JSON] TodoId
 :<|> "todo" :> Capture "id" TodoId :> Get '[JSON] Todo
 :<|> "todo" :> Capture "id" TodoId :> ReqBody '[JSON] Todo :> Put '[JSON] TodoId
```

where the JSON excerpt is the parameters for the "foo" path which I added to the example.

**Comment by fizruk** [comment]
Request and response bodies use `Schema` while query parameters, captures and headers use `ParamSchema`. The main difference between `Schema` and `ParamSchema` is that `Schema` allows objects and heterogenous lists in addition to what `ParamSchema` allows. That is because `Schema` specifies the format for JSON data while `ParamSchema` specifies format for text-like parameters with almost no structure.

So you can't always just copy the `Schema` to `ParamSchema`.
We've worked hard to add a lot of instances and support as many things as we could with Generic-based derivation mechanism, so I now believe that if it does not work for you, it is probably because your data model should be different (e.g. split into pieces) or you should pass your data differently (e.g. use request body instead of query parameters).

Does this explanation work for you?

**Comment by alanz** [comment]
Yes it does, and the key point is that a Schema can be used in the request body, which is what hie does, and explains why the sample works.

Thanks.

**Comment by fizruk** [comment]
Great! I'll keep this issue open as a reminder to include that explanation into the documentation.
